### PR TITLE
[TVM EP] Rename Standalone TVM (STVM) Execution Provider to TVM EP

### PR DESCRIPTION
**Pull request by vvchernov**
Rename all mentions of STVM to TVM in the onnx runtime codebase.

Corrections were done after [PR#10241](https://github.com/microsoft/onnxruntime/pull/10241) and [PR#10211](https://github.com/microsoft/onnxruntime/pull/10211) were merged.

Notes:
1. Nuphar and TVM EPs building were separated in build and cmake files for convenient further work.
2. New key is used for build TVM EP +Cuda to avoid confusion with CUDA EP. Adding the new key is associated with the following `--use_cuda` usage issues:
Using the `--use_cuda flag` means that when building a project, you need to build CUDA EP. This is not the right approach if we only want to work with TVM EP. Building CUDA EP is a separate flow that requires its own nuances and dependencies, which are redundant if we only need TVM EP.
At the `build.py` script level, there are checks for `--cuda_home` and `--cudnn_home` flags, which are not needed for TVM. Also, setting the `--use_cuda` flag turns on a lot of additional logic when running build.py that is not needed for the TVM EP. In order to disable this additional logic, it is necessary to extend conditions if `args.use_cuda`: to if `args.use_cuda` and not `args.use_tvm` in many places. This is not conceptually correct, since these changes for TVM EP need to be made in methods that are specific to CUDA EP.
Additionally, there are problems using TVM EP via PYTHONPATH and the wheel package. This is because setup.py only supports one EP at a time. If we set `--use_cuda` and `--use_tvm` flags, then only one wheel package for CUDA EP will be built, because this is how the logic of working with providers in `setup.py` is arranged. Also, the condition for the `_ld_preload.py` extension for TVM EP will not be fulfilled, with the help of which the
3. If input shapes are unset and dynamic the ONNX runtime error is raised instead of warning and automatical inferred of dynamic dimension to 1.
4. There are two independent structures with the same name (TVMFuncState) in TVM EP internal code and in tests for TVM prepared by NUPHAR.

Hello @xadupre, can you recheck our updates after your PR?

### CODE REVIEW

File: docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb
Comment (xadupre) [inline docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb:318]: Same warning as the removed one but is there any way to remove it?

File: cmake/CMakeLists.txt
Comment (xadupre) [inline cmake/CMakeLists.txt:62]: Why `onnxruntime_NUPHAR_USE_TVM` and not `onnxruntime_USE_NUPHAR_TVM`?

Comment (vvchernov) [reply cmake/CMakeLists.txt:62]: @xadupre it does not metter for me it separated 'USE_TVM' after 'USE_STVM' renaming and 'USE_TVM' used by Nuphar. if you wish I can redo it

Comment (xadupre) [reply cmake/CMakeLists.txt:62]: Every other similar constant starts with `onnxruntime_USE...`. It seems better to follow the same pattern.

Comment (vvchernov) [reply cmake/CMakeLists.txt]: Ok. done

**Review by xadupre — DISMISSED** [review]

**Comment by tmoreau89** [comment]
@xadupre it appears the CI are blocked, would you mind unblocking? thank you.

**Comment by tmoreau89** [comment]
Recommendation is to land this PR after https://github.com/microsoft/onnxruntime/pull/10211 is merged

**Comment by tmoreau89** [comment]
@xadupre and @edgchen1 - we have finished rebasing and fixing some of the build issues we uncovered after https://github.com/microsoft/onnxruntime/pull/10211 merged. Please see the updated PR description.

Would it be possible to re-trigger the CI and get an additional set of reviews? Thanks

File: cmake/CMakeLists.txt
Comment (xadupre) [inline cmake/CMakeLists.txt:1412]: I added this messages to check cmake was doing was I was expected. Feel free to remove them.

Comment (vvchernov) [reply cmake/CMakeLists.txt:1412]: Ok, thanks! @xadupre  Can I remove it without CI test skipping?

File: docs/TVM_EP.md
Comment (xadupre) [inline docs/TVM_EP.md:26]: If TVM has specific requirement on the version, is there a page the documentation can link to?

File: docs/TVM_EP.md
Comment (xadupre) [inline docs/TVM_EP.md:63]: Good catch. linux or os_name is not mandatory. It can be removed.

File: docs/TVM_EP.md
Comment (xadupre) [inline docs/TVM_EP.md:144]: Feel free to remove these lines then.

File: docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb
Comment (xadupre) [inline docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb:314]: About that parameter, I had some issues with it. I had some logs I got on a local machine. I used AutoTVM instead of Ansor and got an error I could not correlate to the value of this parameter. Is there a page on TVM documentation we can refer to?

File: tools/ci_build/build.py
Comment (xadupre) [inline tools/ci_build/build.py:768]: The default configuration can be different of CPU and CUDA. The most efficient should be chosen.

File: cmake/CMakeLists.txt
Comment (xadupre) [reply cmake/CMakeLists.txt:1412]: I meant the following messages. If you remove set(use_openmp..., it will be set for every package built after this one. That's why I did it that way. But I don't know the best setting. It could be set also with the same value as onnxruntime_USE_OPENMP. If it is better without, then let's remove it.

File: tools/ci_build/build.py
Comment (vvchernov) [reply tools/ci_build/build.py:768]: It was TODO for check of possible building pipeline. I discussed it with my colleagues and LLVM is needed in all cases. TODO was removed.

File: docs/TVM_EP.md
Comment (vvchernov) [reply docs/TVM_EP.md:144]: Thanks! I've removed it.

File: docs/TVM_EP.md
Comment (vvchernov) [reply docs/TVM_EP.md:63]: We assume that ONNX runtime is built by build.sh (see TVM_EP.md). Inside the latter the path with OS_DIR is constructed due to this linux or os_name is mandatory

File: docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb
Comment (vvchernov) [reply docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb:314]: TVM has separated documentation for AutoTVM and AutoScheduler (Ansor) using. I've not seen any for issues related to using one instead of another, but I can say that it can occur due to there are different parser for tuning logs from AutoTVM or AutoScheduler. I think we will extend the current documentation by description devoted to using different optimization system but not now. As I know at the moment we have agreement that AutoTVM is supported on ONNX runtime side, therefore it was set as default. Ansor still has some issues and 'tuning type' is developer's key.  @tmoreau89 please confirm or correct me.

File: docs/TVM_EP.md
Comment (KJlaccHoeUM9l) [reply docs/TVM_EP.md]: Hello @xadupre!
There are no special version requirements. 
There are also no additional links to documentation. CUDA is an external dependency in this context, so it is up to the user to decide how they will deliver this dependency.

Comment (tmoreau89) [reply docs/TVM_EP.md]: Internally we're using CUDA 10, but moving over to CUDA 11.1 and ensuring this leads to no performance regressions.

File: docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb
Comment (tmoreau89) [reply docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb:314]: I think the ideal solution here is to derive from the logs what kind of tuner is in use; Ansor or AutoTVM. It should be doable so that user-induced errors don't occur. I think we can address this in a separate PR.

**Comment by vvchernov** [comment]
Hello @xadupre , could you recheck our changes after the last review and restart CI tests?

File: docs/TVM_EP.md
Comment (xadupre) [inline docs/TVM_EP.md:72]: python3

**Comment by vvchernov** [comment]
Hello @xadupre! I've investigated two failed CI tests (Win CPU build_x64_release), they have the same issue:
```
2022-02-09T16:08:56.5210498Z    126>CustomBuild:
2022-02-09T16:08:56.5211264Z          
2022-02-09T16:08:56.5211601Z          FAILURE: Build failed with an exception.
2022-02-09T16:08:56.5213510Z          
2022-02-09T16:08:56.5214425Z          * Where:
2022-02-09T16:08:56.5215726Z          Build file 'D:\a\_work\2\s\java\build.gradle' line: 6
2022-02-09T16:08:56.5217069Z          
2022-02-09T16:08:56.5218173Z          * What went wrong:
2022-02-09T16:08:56.5219477Z          Plugin [id: 'com.diffplug.spotless', version: '5.17.0'] was not found in any of the following sources:
2022-02-09T16:08:56.5221353Z          
2022-02-09T16:08:56.5222786Z          - Gradle Core Plugins (plugin is not in 'org.gradle' namespace)
2022-02-09T16:08:56.5224605Z          - Plugin Repositories (could not resolve plugin artifact 'com.diffplug.spotless:com.diffplug.spotless.gradle.plugin:5.17.0')
2022-02-09T16:08:56.5225990Z            Searched in the following repositories:
2022-02-09T16:08:56.5227564Z              Gradle Central Plugin Repository
2022-02-09T16:08:56.5228922Z          
2022-02-09T16:08:56.5230389Z          * Try:
2022-02-09T16:08:56.5232176Z          > Run with --stacktrace option to get the stack trace.
2022-02-09T16:08:56.5233527Z          > Run with --info or --debug option to get more log output.
2022-02-09T16:08:56.5234239Z          > Run with --scan to get full insights.
2022-02-09T16:08:56.5235438Z          
2022-02-09T16:08:56.5237340Z          * Get more help at https://help.gradle.org
2022-02-09T16:08:56.5238007Z          
2022-02-09T16:08:56.5239669Z          BUILD FAILED in 2m 39s
2022-02-09T16:08:57.5246504Z ##[error]C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(241,5): Error MSB8066: Custom build for 'D:\a\_work\2\b\RelWithDebInfo\CMakeFiles\8e4b62db2edc18ee62160ab55e5804a5\onnxruntime.jar.rule;D:\a\_work\2\b\RelWithDebInfo\CMakeFiles\ecdb4e27a46a08701e25f6bac4ec0295\onnxruntime4j.rule;D:\a\_work\2\s\cmake\CMakeLists.txt' exited with code 1.
```
It looks like it is not related to our updates, particularly the builds do not use `--use_tvm` key. I disscused it with my colleagues and we decided that it can be flaky test or problems on onnx runtime side. Due to the latter I rebased the branch onto current master. Please start Win CI test to quickly check that it works.
If you have other ideas why did it occur please tell us.

**Comment by vvchernov** [comment]
@xadupre It looks like previous failed tests passed successfully. Linux CI tests also can be started

File: docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb
Comment (tmoreau89) [inline docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb:308]: @KJlaccHoeUM9l - this should be cleaned up to just "TVM EP"!

File: docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb
Comment (tmoreau89) [inline docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb:368]: @KJlaccHoeUM9l This should be cleaned up!

File: docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb
Comment (tmoreau89) [inline docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb:451]: @KJlaccHoeUM9l This should be cleaned up!

**Comment by tmoreau89** [comment]
@xadupre I realize that some lines didn't pass the quality check, so I'll go ahead and fix those lines and let you know when CI can be re-triggered. It'll take 5 mins.

**Comment by xadupre** [comment]
> @xadupre I realize that some lines didn't pass the quality check, so I'll go ahead and fix those lines and let you know when CI can be re-triggered. It'll take 5 mins.

I'll trigger that in an hour or so. I don't know if you can trigger them on your own. But that's worth a try. You need to add a new comment with the following line for linux
```
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux GPU CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux Nuphar CI Pipeline, Linux OpenVINO CI Pipeline, MacOS CI Pipeline, ONNX Runtime Web CI Pipeline, onnxruntime-binary-size-checks-ci-pipeline
```
and this one on Windows:
```
/azp run Windows CPU CI Pipeline, Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, Windows WebAssembly CI Pipeline, orttraining-amd-gpu-ci-pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed, onnxruntime-python-checks-ci-pipeline
```

**Comment by tmoreau89** [comment]
Hi Xavier, it would be great to be able to trigger them. I can try.

**Comment by tmoreau89** [comment]
@xadupre it looks like I wasn't able to trigger the test! It would be great to have the privilege, that could help speedup things on our end. The cleanup was done, you can re-trigger whenever you have the time. Thank you!

File: docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb
Comment (tmoreau89) [reply docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb]: resolved

File: docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb
Comment (tmoreau89) [reply docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb]: I made the cleanup; I'll work with Xavier to get this merged, thanks Valery

Comment (vvchernov) [reply docs/python/inference/notebooks/onnxruntime-tvm-tutorial.ipynb]: I saw. It is pitty that trick from Xavier does not work. I'm not sure it looks like the text from ipynb is autogenerated, I asked Alexey to rechecked it. Let's merge this PR, if we find some issue related to the renaming we will fix it further

**Comment by vvchernov** [comment]
Hello @xadupre! I think we need your or somebody from MS help with CI tests. As you could see they failed due to different reasons in different CI tests and there are no obvious links with our code. For example, first time it failed due to `spotless, version: '5.17.0' was not found` in java build in Win CPU CI. When I rebased it onto master all test passed successfully. After Theirry's fixes in comments it failed in two Win GPU CI tests, the issue was related to test of specific op. Just now (after CI tests restart) it again has problem in Win CPU CI, but other reason: C# test failed. TVM EP does not have any C# test.
We stop our development for this PR and can not fix this floating issues due to they are not related to TVM EP. I see two ways: 1. somebody helps us with CI tests 2. you will restart the tests several times while they pass succesfully. For the latter case I assume that it is flaky tests and they should pass as result.

**Comment by xadupre** [comment]
I think PR #10524 is fixing this issue.

**Comment by vvchernov** [comment]
Thanks @xadupre , I wait for PR #10524 will be merged and after that rebase our branch onto master

**Comment by vvchernov** [comment]
Good evening @xadupre. PR #10524 was merged, I rebased our branch onto master, please start CI tests

**Comment by vvchernov** [comment]
PR #10534 and #10533 have the same CI failure. Waiting for fix by somebody

**Comment by tmoreau89** [comment]
@xadupre - is the issue found in `orttraining-amd-gpu-ci-pipeline` a known issue? Currently blocked on getting the CI to go green; seems unrelated to this PR.

**Review by KJlaccHoeUM9l — APPROVED** [review]

**Review by KJlaccHoeUM9l — APPROVED** [review]

**Comment by vvchernov** [comment]
Hello @xadupre! Please check last time and approve the PR

**Review by xadupre — APPROVED** [review]

**Pull request merged**
